src/errors.rs

```diff
@@ -231,3 +231,23 @@ impl Termination for BinstallError {
         code
     }
 }
+
+impl From<std::io::Error> for BinstallError {
+    fn from(err: std::io::Error) -> Self {
+        let is_inner_binstall_err = err
+            .get_ref()
+            .map(|e| e.is::<BinstallError>())
+            .unwrap_or_default();
+
+        if is_inner_binstall_err {
+            let inner = err
+                .into_inner()
+                .expect("err.get_ref() returns Some, so err.into_inner() should als return Some");
+            *inner
+                .downcast()
+                .expect("The inner err is tested to be BinstallError")
```

Comment:
I think this should be doable without `expect`ing at all, I'll try later

Reply:
I found the ways to avoid the second `expect`, so now it's a little bit better.

I don't think there is anyway to avoid `expect`/`unwrap` at all unless `std::io::Error` provides an api like this:

```
fn into_underlying_err(self) -> Cause;
fn from_underlying_err(cause: Cause) -> Self;

#[non_exhaustive]
enum Cause {
    RawOsError(i32),
    InnerError(Box<dyn Error + Send + Sync>),
}
```

Existing APIs requires two separate calls to obtain the raw os error or the inner error and they both return `Option`.
There's no way to completely avoid `unwrap`/`expect` if we want to cover every possible corner case here, e.g., both of them returns `None`.

This is impossible without `expect`/`unwrap` because `Error::into_inner` takes the error by value and returns `Option` instead of `Result<Box<dyn Error + Send + Sync>, Self>`.

One way to fix this situation is to add another variant to `BinstallError` specifically for this case.

Reply:
Actually, I think this is worth a PR to rust-lang/rust.

Reply:
I made a new proposal to rust https://github.com/rust-lang/libs-team/issues/57

P.S. This is my first time doing these kind of stuff